File: modules/action/src/main/scala/action/sourcing/SourcingStream.scala
package action.sourcing
import java.util.UUID

import action.common.{ActionProducer, IdempotentStream, Utils}
import io.simplesource.api.CommandError
import io.simplesource.data.{Result, Sequence}
import io.simplesource.kafka.model.{CommandRequest, CommandResponse}
import model.messages.{ActionRequest, ActionResponse}
import model.saga.SagaError
import org.apache.kafka.common.utils.Bytes
import org.apache.kafka.streams.kstream._
import org.apache.kafka.streams.state.KeyValueStore
import org.slf4j.LoggerFactory
import shared.streams.syntax._

object SourcingStream {
  private val logger = LoggerFactory.getLogger("SourcingStream")

  def logValues[K, V](prefix: String): ForeachAction[K, V] = Utils.logValues[K, V](logger, prefix)

  def addSubTopology[A, I, K, C](ctx: SourcingContext[A, I, K, C],
                                 actionRequest: KStream[UUID, ActionRequest[A]],
                                 actionResponse: KStream[UUID, ActionResponse],
                                 commandResponseByAggregate: KStream[K, CommandResponse]): Unit = {
    val commandResponseByCommandId = commandResponseByAggregate.selectKey[UUID]((_, v) => v.commandId())

    // join the action request with corresponding prior command responses
    val idempotentAction =
      IdempotentStream.getActionRequestsWithResponse(ctx.actionSpec,
                                                     actionRequest,
                                                     actionResponse,
                                                     ctx.commandSpec.actionType)

    // get new command requests
    val (requestErrorResponses, commandRequests) =
      handleActionRequest(ctx, idempotentAction.unprocessedRequests, commandResponseByAggregate)

    // handle incoming command responses
    val newActionResponses = handleCommandResponse(ctx, actionRequest, commandResponseByCommandId)

    // publish to output topics
    CommandProducer.commandRequest(ctx.commandSpec, commandRequests)
    ActionProducer.actionResponse(ctx.actionSpec,
                                  idempotentAction.priorResponses,
                                  newActionResponses,
                                  requestErrorResponses)
  }

  def handleActionRequest[A, I, K, C](ctx: SourcingContext[A, I, K, C],
                                      actionRequests: KStream[UUID, ActionRequest[A]],
                                      commandResponseByAggregate: KStream[K, CommandResponse])
    : (KStream[UUID, ActionResponse], KStream[K, CommandRequest[K, C]]) = {

    def getAggregateSequence(cResp: CommandResponse): Sequence =
      cResp.sequenceResult().getOrElse(cResp.readSequence())

    val reqsWithDecoded: KStream[UUID, (ActionRequest[A], Either[Throwable, I])] =
      actionRequests._mapValues(ar => (ar, ctx.commandSpec.decode(ar.actionCommand.command)))

    val errorActionRequests = reqsWithDecoded
      ._collect {
        case (ard, Left(e)) =>
          ActionResponse(sagaId = ard.sagaId,
                         actionId = ard.actionId,
                         commandId = ard.actionCommand.commandId,
                         Left(SagaError.of(e.getMessage)))
      }

    val allGood = reqsWithDecoded._collect { case (ar, Right(i)) => (ar, i) }

    // Sort incoming request by the aggregate key
    val requestByAggregateKey: KStream[K, ActionRequest[A]] = allGood
      ._map((_, aReq: (ActionRequest[A], I)) => (ctx.commandSpec.keyMapper(aReq._2), aReq._1))
      .peek(logValues[K, ActionRequest[A]]("requestByAggregateKey"))

    // Get the most recent command response for the aggregate
    val lastCommandByAggregate: KTable[K, CommandResponse] = {
      val materializer: Materialized[K, CommandResponse, KeyValueStore[Bytes, Array[Byte]]] =
        Materialized
          .as[K, CommandResponse, KeyValueStore[Bytes, Array[Byte]]](
            "last_command_by_aggregate_" + ctx.commandSpec.aggregateName)
          .withKeySerde(ctx.cSerdes.aggregateKey)
          .withValueSerde(ctx.cSerdes.response)

      commandResponseByAggregate
        .groupByKey()
        .reduce((cr1: CommandResponse, cr2: CommandResponse) =>
                  if (getAggregateSequence(cr2).isGreaterThan(getAggregateSequence(cr1))) cr2
                  else cr1,
                materializer)
    }

    // Get the latest sequence number and turn action request into a command request
    val commandRequestByAggregate: KStream[K, CommandRequest[K, C]] = {
      val valueJoiner: ValueJoiner[ActionRequest[A], CommandResponse, CommandRequest[K, C]] =
        (aReq: ActionRequest[A], cResp: CommandResponse) => {
          val sequence =
            if (cResp == null) Sequence.first() else getAggregateSequence(cResp)
          // we can do this safely as we have (unfortunately) already done this, and succeeded first time
          val intermediate = ctx.commandSpec.decode(aReq.actionCommand.command).right.get
          new CommandRequest[K, C](ctx.commandSpec.keyMapper(intermediate),
                                   ctx.commandSpec.commandMapper(intermediate),
                                   sequence,
                                   aReq.actionCommand.commandId)
        }

      requestByAggregateKey.leftJoin[CommandResponse, CommandRequest[K, C]](
        lastCommandByAggregate,
        valueJoiner,
        Joined.`with`[K, ActionRequest[A], CommandResponse](ctx.cSerdes.aggregateKey,
                                                            ctx.aSerdes.request,
                                                            ctx.cSerdes.response)
      )
    }.peek(logValues[K, CommandRequest[K, C]]("commandRequestByAggregate"))

    (errorActionRequests, commandRequestByAggregate)
  }

  def handleCommandResponse[A, I, K, C](
      ctx: SourcingContext[A, I, K, C],
      actionRequests: KStream[UUID, ActionRequest[A]],
      responseByCommandId: KStream[UUID, CommandResponse]): KStream[UUID, ActionResponse] = {
    // find the response for the request
    val actionRequestWithResponse: KStream[UUID, (ActionRequest[A], CommandResponse)] = {
      val valueJoiner: ValueJoiner[ActionRequest[A], CommandResponse, (ActionRequest[A], CommandResponse)] =
        (aReq: ActionRequest[A], cResp: CommandResponse) => (aReq, cResp)

      // join command response to action request by the command / action ID
      // TODO: timeouts - will be easy to do timeouts with a left join once https://issues.apache.org/jira/browse/KAFKA-6556 has been released
      val timeOutMillis = ctx.commandSpec.timeOutMillis
      actionRequests
        .selectKey[UUID]((_: UUID, aReq: ActionRequest[A]) => aReq.actionCommand.commandId)
        .join[CommandResponse, (ActionRequest[A], CommandResponse)](
          responseByCommandId,
          valueJoiner,
          JoinWindows.of(timeOutMillis).until(timeOutMillis * 2 + 1),
          Joined.`with`(ctx.aSerdes.uuid, ctx.aSerdes.request, ctx.cSerdes.response)
        )
        .peek(logValues[UUID, (ActionRequest[A], CommandResponse)]("handleCommandResponse"))

    }

    // turn the pair into an ActionResponse
    getActionResponse(ctx, actionRequestWithResponse)
  }

  def getActionResponse[A, I, K, C](
      ctx: SourcingContext[A, I, K, C],
      actionRequestWithResponse: KStream[UUID, (ActionRequest[A], CommandResponse)])
    : KStream[UUID, ActionResponse] = {
    val actionResponse: KStream[UUID, ActionResponse] = actionRequestWithResponse
      ._mapValues[ActionResponse] {
        case (aReq, cResp) =>
          val sequenceResult: Result[CommandError, Sequence] =
            if (cResp == null)
              Result.failure(CommandError.of(CommandError.Reason.Timeout,
                                             "Timed out waiting for response from Command Processor"))
            else
              cResp.sequenceResult()

          import scala.collection.JavaConverters._
          val result =
            sequenceResult
              .fold[Either[SagaError, Unit]](
                nelE => Left(SagaError.of(nelE.head().getMessage, nelE.tail().asScala.map(_.getMessage): _*)),
                _ => Right(()))

          ActionResponse(sagaId = aReq.sagaId,
                         actionId = aReq.actionId,
                         commandId = aReq.actionCommand.commandId,
                         result = result)
      }
      .selectKey[UUID]((_: UUID, sagaId: ActionResponse) => sagaId.sagaId)
      .peek(logValues[UUID, ActionResponse]("resultStream"))

    actionResponse
  }
}

File: modules/action/src/main/java/io/simplesource/saga/action/sourcing/SourcingStream.java
package io.simplesource.saga.action.sourcing;

import io.simplesource.api.CommandError;
import io.simplesource.data.NonEmptyList;
import io.simplesource.data.Result;
import io.simplesource.data.Sequence;
import io.simplesource.kafka.internal.util.Tuple2;
import io.simplesource.kafka.model.CommandRequest;
import io.simplesource.kafka.model.CommandResponse;
import io.simplesource.saga.action.common.ActionProducer;
import io.simplesource.saga.action.common.IdempotentStream;
import io.simplesource.saga.action.common.Utils;
import io.simplesource.saga.model.messages.ActionRequest;
import io.simplesource.saga.model.messages.ActionResponse;
import io.simplesource.saga.model.saga.SagaError;
import org.apache.kafka.common.utils.Bytes;
import org.apache.kafka.streams.KeyValue;
import org.apache.kafka.streams.kstream.*;
import org.apache.kafka.streams.state.KeyValueStore;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.UUID;
import java.util.function.Function;


public class SourcingStream {
    static Logger logger = LoggerFactory.getLogger(SourcingStream.class);

    public static <K, V> ForeachAction<K, V> logValues(String prefix) {
        return (k, v) -> Utils.logValues(logger, prefix);
    }

    public static <A, I, K, C> void addSubTopology(SourcingContext<A, I, K, C> ctx,
                                                   KStream<UUID, ActionRequest<A>> actionRequest,
                                                   KStream<UUID, ActionResponse> actionResponse,
                                                   KStream<K, CommandResponse> commandResponseByAggregate) {
        KStream<UUID, CommandResponse> commandResponseByCommandId = commandResponseByAggregate.selectKey((k, v) -> v.commandId());

        IdempotentStream.IdempotentAction<A> idempotentAction = IdempotentStream.getActionRequestsWithResponse(ctx.actionSpec,
                actionRequest,
                actionResponse,
                ctx.commandSpec.actionType);
        // get new command requests
        Tuple2<KStream<UUID, ActionResponse>, KStream<K, CommandRequest<K, C>>> requestResp = handleActionRequest(ctx, idempotentAction.unprocessedRequests, commandResponseByAggregate);
        KStream<UUID, ActionResponse> requestErrorResponses = requestResp.v1();
        KStream<K, CommandRequest<K, C>> commandRequests = requestResp.v2();

        // handle incoming command responses
        KStream<UUID, ActionResponse> newActionResponses = handleCommandResponse(ctx, actionRequest, commandResponseByCommandId);

        // publish to output topics
        CommandProducer.commandRequest(ctx.commandSpec, ctx.commandTopicNamer, commandRequests);
        ActionProducer.actionResponse(ctx.actionSpec,
                ctx.actionTopicNamer,
                idempotentAction.priorResponses,
                newActionResponses,
                requestErrorResponses);
    }


    public static <A, I, K, C>
    Tuple2<KStream<UUID, ActionResponse>, KStream<K, CommandRequest<K, C>>> handleActionRequest(
            SourcingContext<A, I, K, C> ctx,
            KStream<UUID, ActionRequest<A>> actionRequests,
            KStream<K, CommandResponse> commandResponseByAggregate) {

        Function<CommandResponse, Sequence> getAggregateSequence = cResp ->
                cResp.sequenceResult().getOrElse(cResp.readSequence());

        KStream<UUID, Tuple2<ActionRequest<A>, Result<Throwable, I>>> reqsWithDecoded =
                actionRequests.mapValues((k, ar) -> Tuple2.of(ar, ctx.commandSpec.decode.apply(ar.actionCommand.command)));

        KStream<UUID, Tuple2<ActionRequest<A>, Result<Throwable, I>>>[] branchSuccessFailure = reqsWithDecoded.branch((k, v) -> v.v2().isSuccess(), (k, v) -> v.v2().isFailure());

        KStream<UUID, ActionResponse> errorActionResponses = branchSuccessFailure[1].mapValues((k, v) -> {
            ActionRequest<A> request = v.v1();
            NonEmptyList<Throwable> reasons = v.v2().failureReasons().get();
            return new ActionResponse(request.sagaId, request.actionId, request.actionCommand.commandId, Result.failure(
                    SagaError.of(SagaError.Reason.InternalError, reasons.head())));
        });

        KStream<UUID, Tuple2<ActionRequest<A>, I>> allGood = reqsWithDecoded.mapValues((k, v) -> Tuple2.of(v.v1(), v.v2().getOrElse(null)));

        // Sort incoming request by the aggregate key
        KStream<K, ActionRequest<A>> requestByAggregateKey = allGood
                .map((k, v) -> KeyValue.pair(ctx.commandSpec.keyMapper.apply(v.v2()), v.v1()))
                .peek(logValues("requestByAggregateKey"));

        Materialized<K, CommandResponse, KeyValueStore<Bytes, byte[]>> materializer =
                Materialized
                        .<K, CommandResponse, KeyValueStore<Bytes, byte[]>>as(
                                "last_command_by_aggregate_" + ctx.commandSpec.aggregateName)
                        .withKeySerde(ctx.cSerdes().aggregateKey())
                        .withValueSerde(ctx.cSerdes().commandResponse());

        // Get the most recent command response for the aggregate
        KTable<K, CommandResponse> lastCommandByAggregate =
                commandResponseByAggregate
                        .groupByKey()
                        .reduce((cr1, cr2) ->
                                        getAggregateSequence.apply(cr2).isGreaterThan(getAggregateSequence.apply(cr1)) ? cr2 : cr1,
                                materializer);

        ValueJoiner<ActionRequest<A>, CommandResponse, CommandRequest<K, C>> valueJoiner =
                (aReq, cResp) -> {
                    Sequence sequence = (cResp == null) ? Sequence.first() : getAggregateSequence.apply(cResp);

                    // we can do this safely as we have (unfortunately) already done this, and succeeded first time
                    I intermediate = ctx.commandSpec.decode.apply(aReq.actionCommand.command).getOrElse(null);
                    return new CommandRequest<>(ctx.commandSpec.keyMapper.apply(intermediate),
                            ctx.commandSpec.commandMapper.apply(intermediate),
                            sequence,
                            aReq.actionCommand.commandId);
                };

        // Get the latest sequence number and turn action request into a command request
        KStream<K, CommandRequest<K, C>> commandRequestByAggregate = requestByAggregateKey
                .leftJoin(
                        lastCommandByAggregate,
                        valueJoiner,
                        Joined.with(ctx.cSerdes().aggregateKey(),
                                ctx.aSerdes().request(),
                                ctx.cSerdes().commandResponse()))
                .peek(logValues("commandRequestByAggregate"));

        return Tuple2.of(errorActionResponses, commandRequestByAggregate);
    }


    public static <A, I, K, C> KStream<UUID, ActionResponse> handleCommandResponse(
            SourcingContext<A, I, K, C> ctx,
            KStream<UUID, ActionRequest<A>> actionRequests,
            KStream<UUID, CommandResponse> responseByCommandId) {
        long timeOutMillis = ctx.commandSpec.timeOutMillis;
        // find the response for the request
        KStream<UUID, Tuple2<ActionRequest<A>, CommandResponse>> actionRequestWithResponse =

                // join command response to action request by the command / action ID
                // TODO: timeouts - will be easy to do timeouts with a left join once https://issues.apache.org/jira/browse/KAFKA-6556 has been released
                actionRequests
                        .selectKey((k, aReq) -> aReq.actionCommand.commandId)
                        .join(
                                responseByCommandId,
                                Tuple2::of,
                                JoinWindows.of(timeOutMillis).until(timeOutMillis * 2 + 1),
                                Joined.with(ctx.aSerdes().uuid(), ctx.aSerdes().request(), ctx.cSerdes().commandResponse())
                        )
                        .peek(logValues("handleCommandResponse"));


        // turn the pair into an ActionResponse
        return getActionResponse(ctx, actionRequestWithResponse);

    }

    static <A, I, K, C> KStream<UUID, ActionResponse> getActionResponse(
            SourcingContext<A, I, K, C> ctx,
            KStream<UUID, Tuple2<ActionRequest<A>, CommandResponse>> actionRequestWithResponse) {

        return actionRequestWithResponse
                .mapValues((k, v) -> {
                            ActionRequest<A> aReq = v.v1();
                            CommandResponse cResp = v.v2();
                            Result<CommandError, Sequence> sequenceResult =
                                    (cResp == null) ?
                                            Result.failure(CommandError.of(CommandError.Reason.Timeout,
                                                    "Timed out waiting for response from Command Processor")) :
                                            cResp.sequenceResult();
                            Result<SagaError, Boolean> result =
                                    sequenceResult.fold(errors -> {
                                        String message = String.join(",", errors.map(CommandError::getMessage));
                                                return Result.failure(SagaError.of(SagaError.Reason.CommandError, message));
                                            },
                                            seq -> Result.success(true));

                            return new ActionResponse(aReq.sagaId,
                                    aReq.actionId,
                                    aReq.actionCommand.commandId,
                                    result);
                        }
                )
                .selectKey((k, resp) -> resp.sagaId)
                .peek(logValues("resultStream"));
    }

}
